hypothesis-python/src/hypothesis/core.py

```diff
@@ -1086,6 +1136,9 @@ def _execute_once_for_engine(self, data: ConjectureData) -> None:
                 # This can happen if an error occurred in a finally
                 # block somewhere, suppressing our original StopTest.
                 # We raise a new one here to resume normal operation.
+                # TODO: this should maybe inspect the stacktrace to see that the above
+                # mentioned story is true? I.e. reraise as StopTest iff there is a
+                # StopTest somewhere in the tree of e.__context__
```

Comment:
that would only override the `__cause__` no? And the comment specifically talks about errors happening in `finally`, where you rarely raise explicit exceptions.
It seems pretty dangerous to ignore all exceptions in case the data is frozen, shouldn't the user be informed if there's errors happening in their teardown code?
Although thinking about it more, the problem is perhaps that the `StopTest.__cause__` is not getting reraised/logged?

Reply:
Yeah, I think you're correct here, we do want to expose errors in teardown code.

> Although thinking about it more, the problem is perhaps that the `StopTest.__cause__` is not getting reraised/logged?

hmm... we never `raise StopTest(...) from ...`, but we certainly do raise it from inside exception handlers that we don't want to expose the user to (via something like `except ValueError: data.mark_invalid()`).  I guess we could inspect `__cause__`, but it'd have so many missed alarms that really doesn't seem worth doing :/

Reply:
> hmm... we never raise StopTest(...) from ...

you do it on the very next line in this file :sweat_smile: 
```python
                 raise StopTest(data.testcounter) from e
```
Though I suppose saving it as `__cause__` for reraising in the engine is perhaps not the way to do it, and the caught exception should instead be saved Somewhere Else:tm:

Reply:
Oh and to be explicit, I think traversing `__context__` for `StopTest` would be robust. As far as I know there's no standard way of losing entries in it - but not terribly elegant to traverse when you also have to deal with exception groups.

I'll try to write a test and open a separate issue/PR for it

Reply:
opened #4132 to continue the discussion